Validate values passed for 'unknown'

sirosen opened the pull request:
There are three places that a user can pass a value for 'unknown':
- `Meta` in a Schema class definition
- `Schema.__init__`
- `Schema.load`

For each of these, call an internal check to confirm that the value is valid. In cases where `unknown or self.unknown` is used to support `None`, do the validation after the `or`.

A new test case covers these three locations and confirms that passing an unknown (hah!) value causes a ValueError.

closes #1721, #1732

sirosen committed:
Validate values passed for 'unknown'

There are three places that a user can pass a value for 'unknown':
- Meta in a Schema class definition
- Schema.__init__
- Schema.load

For each of these, call an internal check to confirm that the value is
valid. In cases where `unknown or self.unknown` is used to support
`None`, do the validation after the `or`.

A new test case covers these three locations and confirms that passing
an unknown (hah!) value causes a ValueError.

closes #1721, #1732

lafrech commented:
Thanks for this.

> In cases where unknown or self.unknown is used to support None, do the validation after the or.

I suppose it would be slightly more efficient to only validate `unknown` in this case as `self.unknown` is supposed to be valid since it was validated earlier (thanks to this PR).

The code would be a little longer, though. Not a one-liner anymore.

Not sure this is relevant. The validation itself is just one statement after all:

```py
    if obj not in (INCLUDE, EXCLUDE, RAISE):
```

(Would a (precomputed at import) `set` be marginally faster?)

sirosen commented:
> I suppose it would be slightly more efficient to only validate `unknown` in this case as `self.unknown` is supposed to be valid since it was validated earlier (thanks to this PR).
> 
> The code would be a little longer, though. Not a one-liner anymore.

I was thinking the same thing and then I realized that there's a nice way to write this:
```python
unknown = self.unknown if unknown is None else validate_unknown_parameter_value(unknown)
```

I'll make that change since it should be slightly faster when `load` is called without the `unknown` arg (which is where I think it really matters to try to make things fast).

I don't think it really matters -- once these changes are tweaked as above -- to use a module-scoped set, but I'll make that change too. We're talking about it and I see no reason _not_ to do it that way.

sirosen committed:
Optimize validation of the `unknown` parameter

Avoid calling the validator if a value which is already validated is
being used by doing an `if unknown is None` check. Use a module-scoped
(i.e. global) set in the validation helper to make it slightly faster.

sloria committed:
Simplify assertion

sloria committed:
Merge branch 'dev' into validate-include-param

sloria committed:
Update changelog
